R/env.R

```diff
@@ -311,176 +367,220 @@ mut_parent_env <- function(env, new_env) {
 }
 
 
-#' Assign objects to an environment.
-#'
-#' These functions create bindings in the specified environment. The
-#' bindings are supplied as pairs of names and values, either directly
-#' (`env_assign()`), in dots (`env_define()`), or from a dictionary
-#' (`env_bind()`). See [is_dictionaryish()] for the definition of a
-#' dictionary.
-#'
-#' These functions operate by side effect. For example, if you assign
-#' bindings to a closure function, the environment of the function is
-#' modified in place.
-#'
-#' @inheritParams get_env
-#' @param nm The name of the binding.
-#' @param x The value of the binding.
-#' @param ... Pairs of unique names and R objects used to define new
-#'   bindings.
-#' @return The input object `env`, with its associated environment
+#' Bind symbols to objects in an environment
+#'
+#' @description
+#'
+#' These functions create bindings in an environment. These bindings
+#' are supplied through `...` as pairs of names and values or
+#' expressions. `env_bind()` is equivalent to evaluating a `<-`
+#' expression within the given environment, e.g. `with_env(env, foo <-
+#' "bar")`. This function should take care of the majority of use
+#' cases but the other variants can be useful for specific problems.
+#'
+#' - `env_bind()` takes named _values_. The arguments are evaluated
+#'   once (with [explicit splicing][dots_list]) and bound in `.env`.
+#'
+#' - `env_bind_fns()` takes named _functions_ and creates active
+#'   bindings in `.env`. An active binding executes a function each
+#'   time it is evaluated. `env_bind_fns()` takes dots with [implicit
+#'   splicing][dots_splice], so that you can supply both named
+#'   functions and named lists of functions.
+#'
+#'   If these functions are [closures][is_closure] they are lexically
+#'   scoped in the environment that they bundle. These functions can
+#'   thus refer to symbols from this enclosure that are not actually
+#'   in scope in the dynamic environment where the active bindings are
+#'   invoked. This allows creative solutions to difficult problems
+#'   (see the implementations of `dplyr::do()` methods for an
+#'   example).
+#'
+#' - `env_bind_exprs()` takes named _expressions_. The arguments are
+#'   captured with [exprs()] (and thus support call-splicing and
+#'   unquoting) and assigned to symbols in `.env`. These expressions
+#'   are not evaluated immediately but lazily. Once a symbol is
+#'   evaluated, the corresponding expression is evaluated in turn and
+#'   its value is bound to the symbol (the expressions are thus
+#'   evaluated only once, if at all).
+#'
+#' @details
+#'
+#' Since environments have reference semantics (see relevant section
+#' in [env()] documentation), modifying the bindings of an environment
+#' produces effects in all other references to that environment. In
+#' other words, `env_bind()` and its variants have side effects.
```

Comment:
Maybe this should be `@section Side effects` and since it's called primarily for its side-effects, it should return the first arg invisibly.